src/coprocessor/dag/batch_executor/scan_executor.rs

```diff
@@ -0,0 +1,237 @@
+            if let Some((key, value)) = some_row {
+                // Retrieved one row from point range or non-point range.
+                self.scanned_rows_per_range
+                    .last_mut()
+                    .map_or((), |val| *val += 1);
+
+                self.imp.process_kv_pair(&key, &value, columns)?;
+
+                if columns.rows_len() >= expect_rows {
+                    return Ok(false); // not drained
+                }
+            } else {
+                // No more row in the range.
+                self.ranges.notify_drained();
+            }
+        }
+    }
+}
+
+/// Extracts `FieldType` from `ColumnInfo`.
+// TODO: Embed FieldType in ColumnInfo directly in Cop DAG v2 to remove this function.
+pub fn field_type_from_column_info(ci: &ColumnInfo) -> FieldType {
+    let mut field_type = FieldType::new();
+    field_type.set_tp(ci.get_tp());
+    field_type.set_flag(ci.get_flag() as u32); // FIXME: This `as u32` is really awful.
+    field_type.set_flen(ci.get_columnLen());
+    field_type.set_decimal(ci.get_decimal());
+    field_type.set_collate(ci.get_collation());
+    // Note: Charset is not provided in column info.
+    field_type
+}
+
+impl<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> BatchExecutor for ScanExecutor<S, I, P> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.imp.schema()
+    }
+
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        assert!(!self.is_ended);
+        assert!(expect_rows > 0);
+
+        let mut data = self.imp.build_column_vec(expect_rows);
+        let is_drained = self.fill_column_vec(expect_rows, &mut data);
+
+        // After calling `fill_column_vec`, columns' length may not be identical when some of the
+        // columns are correctly decoded while others are not. So let's trim columns.
+        {
+            let mut min_len = data.rows_len();
+            for col in data.as_ref() {
+                min_len = min_len.min(col.len());
+            }
+            for col in data.as_mut() {
+                col.truncate(min_len);
+            }
+            data.assert_columns_equal_length();
+        }
+
+        // TODO
+        // If `is_drained.is_err()`, it means that there is an error after *successfully* retrieving
+        // these rows. After that, if we only consumes some of the rows (TopN / Limit), we should
+        // ignore this error.
+
+        match &is_drained {
+            Err(_) => self.is_ended = true,
+            Ok(true) => self.is_ended = true,
+            Ok(false) => {}
+        };
+
+        BatchExecuteResult {
+            data,
+            is_drained,
+            warnings: self.imp.mut_context().take_warnings(),
+        }
+    }
+
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        for (index, num) in self.scanned_rows_per_range.iter_mut().enumerate() {
+            destination.scanned_rows_per_range[index] += *num;
```

Comment:
Mainly because we want incremental statistics instead of absolute values. When these executors are executed in a way that we can freely pause and continue (Q2 goal), incremental statistics looks to be better that it can handle the case that some requests may not be fully processed.

Reply:
I still do not understand.. In my opinion, `scanned_rows_per_range `  collects the number of rows scanned by each range in `tablescan` or `indexscan`, so when we finish a `scan` in one range, the value wouldn't change.. Meanwhile, one range will start to scan unless the ranges in front of it finished, so once a range started, the number of rows scan by the ranges in front of it wouldn't be changed?. Then when will the `add` operation happen?

Reply:
Maybe Chinese is better?

原因是希望对外提供的 statistics 是一种可累积的差值，而不是绝对值。对外提供累积差值的话，在使用的时候，只需提供零值，即可收集到「一段时间内」的 statistics；如果采用绝对值，需要在外侧计算 delta 才能知道一段时间内的 statistics。因此，这里每次 collect_statistics 被调用的时候，都是对外累加差值。当然，累加差值还可以实现为返回一个新结构体，结构体内包含了距离上次 collect_statistics 被调用时候的差值，如果是这样提供接口的话，也是直接赋值了，但考虑到我们的 statistics 结构比较复杂，甚至里面有个 Vec，因此每次都返回新结构体不是一个好主意。

至于为什么外面需要「一段时间内」的 statistics——因为未来这些算子不会被一次性执行完毕，而会执行一部分被挂起再执行另一算子。在这种情况下，算子操作所有执行完再汇报统计信息不太科学，可能会导致超大查询被遗漏（一直都在执行），比较理想的改进就是，每执行完一部分，就把统计信息汇报一下，这样不管请求多大都能有及时的统计信息了。